File: slack-reader/src/main/java/dev/number6/slackreader/model/Message.java
package dev.number6.slackreader.model;

import java.util.Objects;
import java.util.Optional;

public class Message {

    public static final String BOT_MESSAGE_SUBTYPE = "bot_message";
    private final String type;
    private final String subtype;
    private final String text;

    public Message(String type, String text) {
        this(type, text, null);
    }

    public Message(String type, String text, String subtype) {
        this.text = text;
        this.type = type;
        this.subtype = subtype;
    }

    public static Message fromBot(Message message) {
        return new Message(message.getType(), message.getText(), BOT_MESSAGE_SUBTYPE);
    }

    public static Message empty(Message message) {
        return new Message(message.getType(), "");
    }

    public static Message nullText(Message message) {
        return new Message(message.getType(), null);
    }

    public static Message spaceText(Message message) {
        return new Message(message.getType(), "        ");
    }

    public String getType() {
        return type;
    }

    public String getText() {
        return text;
    }

    public Optional<String> getSubtype() {
        return Optional.ofNullable(subtype);
    }

    public boolean isBotMessage() {
        return getSubtype().orElse("").equalsIgnoreCase(Message.BOT_MESSAGE_SUBTYPE);
    }

    @Override
    public String toString() {
        return "Message{" +
                "type='" + type + '\'' +
                ", subtype='" + subtype + '\'' +
                ", text='" + text + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Message message = (Message) o;
        return type.equals(message.type) &&
                Objects.equals(subtype, message.subtype) &&
                text.equals(message.text);
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, subtype, text);
    }
}

File: slack-reader/src/main/kotlin/dev/number6/slackreader/model/Message.kt
package dev.number6.slackreader.model


import java.util.*

class Message @JvmOverloads constructor(val type: String,
                                        val text: String?,
                                        private val subtype: String? = null) {

    private fun getSubtype(): Optional<String> {
        return Optional.ofNullable(subtype)
    }

    fun isBotMessage(): Boolean {
        return getSubtype().orElse("").equals(BOT_MESSAGE_SUBTYPE, ignoreCase = true)
    }

    override fun toString(): String {
        return "Message{" +
                "type='" + type + '\'' +
                ", subtype='" + subtype + '\'' +
                ", text='" + text + '\'' +
                '}'
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val message = o as Message
        return type == message.type &&
                subtype == message.subtype && text == message.text
    }

    override fun hashCode(): Int {
        return Objects.hash(type, subtype, text)
    }

    companion object {
        const val BOT_MESSAGE_SUBTYPE = "bot_message"

        fun fromBot(message: Message): Message {
            return Message(message.type, message.text, BOT_MESSAGE_SUBTYPE)
        }

        fun empty(message: Message): Message {
            return Message(message.type, "")
        }

        fun nullText(message: Message): Message {
            return Message(message.type, null)
        }

        fun spaceText(message: Message): Message? {
            return Message(message.type, "        ")
        }
    }

}